mlflow/server/js/src/common/utils/AuthService.js

```diff
@@ -0,0 +1,91 @@
+export class AuthService {
+  token = null;
+  toRedirectState = '';
+  ssoLoginHeader = 'X-SSO-Login';
+  ssoLogoutHeader = 'X-SSO-Logout';
+  redirectUriRoute = 'oauth';
+
+  constructor() {
+    const savedToken = localStorage.getItem('token');
+    if (savedToken !== null) {
+      this.token = savedToken;
+    }
+  }
+
+  logOff() {
+    this.setNewTokenInternal(null);
+    const fullLogoutUrl = this.getFullLogoutUrl();
+    if (fullLogoutUrl !== null) {
+      window.location.assign(fullLogoutUrl);
+    } else {
+      window.location.reload();
+    }
+  }
+
+  getFullLogoutUrl() {
+    const logoutUrl = localStorage.getItem('logoutUrl');
+    if (logoutUrl !== null) {
+      return (
+        logoutUrl +
+        '&goto=' +
+        encodeURIComponent(window.location.protocol + '//' + window.location.host + '/')
+      );
+    }
+    return null;
+  }
+
+  getToken() {
+    return this.token;
+  }
+
+  setNewToken(token) {
+    this.setNewTokenInternal(token);
+  }
+
+  clearAuth() {
+    this.setNewTokenInternal(null);
+  }
+
+  redirectUrl() {
+    return window.location.protocol + '//' + window.location.host + '/' + this.redirectUriRoute;
+  }
+
+  /**
+   * If can retrieve the SSO Url, intiate a redirect. else print an error
+   * @param unauthorized the unauthorized http response
+   */
+  redirectToSsoIfPossible(unauthorized) {
+    window.location.assign(this.getSsoUrl(unauthorized));
```

Comment:
I don t know how the React router works, but here you can save the current state (everything after the hostname in the url), to restore it after being back from the SSO (and the token retrieved). you can pass a state parameter to the SSO, which will give it you back after the redirect

Reply:
Done